marshal/admin.go

```diff
@@ -0,0 +1,358 @@
+const (
+	consumerReleaseClaimWaitSleep = time.Duration(5 * time.Second)
+	consumerReleaseClaimWaitTime  = 15 * time.Minute
+)
+
+// Admin is used to pause a consumer group and set what position it reads from
+// for certain partitions.
+type Admin interface {
+	// SetConsumerGroupPosition resets consumers to read starting from
+	// offsets on each topic, partition pair in positions.
+	SetConsumerGroupPosition(groupID string, offsets map[string]map[int]int64) error
+}
+
+type consumerGroupAdmin struct {
+	clientID     string
+	groupID      string
+	marshaler    *Marshaler
+	pauseTimeout time.Duration
+
+	// steps is for testing -- it is roughly the number of messages we produce to Marshal.
+	steps *int32
+	// claimHealth is 0 if any of our successfully-claimed claims fail to heartbeat.
+	claimHealth *int32
+	// The lock protects the structs below.
+	lock *sync.RWMutex
+	// claims are partitions that we've successfully claimed after they've been released,
+	// that we'd like to reset the offsets for.
+	claims []claimAttempt
+	// releaseGroupPartitions keeps track of which Marshal partitions
+	// we've produced ReleaseGroup  messages to.
+	releaseGroupPartitions []int32
+}
+
+// claimAttempt represents a topic, partition we'd like to reset the offset of.
+type claimAttempt struct {
+	topic  string
+	partID int
+	// What we'd like to set the offset of a particular partition to be.
+	newOffset int64
+	// What the current offset of a particular partition is.
+	currentOffset int64
+}
+
+// addClaimAttempt adds a successfully-claimed partition to our Admin.
+func (a *consumerGroupAdmin) addClaimAttempt(topic string,
+	partID int, currentOffset, newOffset int64) {
+
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	c := claimAttempt{
+		topic:         topic,
+		partID:        partID,
+		newOffset:     newOffset,
+		currentOffset: currentOffset}
+	a.claims = append(a.claims, c)
+}
+
+// claimHealth returns whether or not any of the admin's claims have failed to heartbeat.
+func (a *consumerGroupAdmin) claimsHealthy() bool {
+	return atomic.LoadInt32(a.claimHealth) == 0
+}
+
+// NewAdmin returns a new Admin struct bound to a Marshaler. The Marshaler should not have
+// any consumers associated with it.
+func (m *Marshaler) NewAdmin(groupID string, pauseTimeout time.Duration) (Admin, error) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	if len(m.consumers) != 0 {
+		return nil, fmt.Errorf(
+			"The Marshaler instance bound to an Admin should not have any consumers.")
+	}
+	return &consumerGroupAdmin{
+		clientID:     m.clientID,
+		groupID:      groupID,
+		marshaler:    m,
+		pauseTimeout: pauseTimeout,
+
+		claimHealth: new(int32),
+		lock:        &sync.RWMutex{},
+		steps:       new(int32),
+	}, nil
+}
+
+// release releases an Admin's claim on a partition. Optionally resets the offset on the partition.
+func (a consumerGroupAdmin) release(topic string, partID int, offset int64, withNewOffset bool) bool {
+	if withNewOffset {
```

Comment:
You shouldn't need to heartbeat before releasing. The ReleasePartition message includes the offset and also calls OffsetCommit so it's the same as heartbeating.